flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RecoveredInputChannel.java

```diff
@@ -177,17 +178,18 @@ void sendTaskEvent(TaskEvent event) {
 
 	@Override
 	boolean isReleased() {
-		return isReleased;
+		return isReleased.get();
 	}
 
 	void releaseAllResources() throws IOException {
-		ArrayDeque<Buffer> releasedBuffers = new ArrayDeque<>();
-		synchronized (receivedBuffers) {
-			releasedBuffers.addAll(receivedBuffers);
-			receivedBuffers.clear();
-			isReleased = true;
+		if (isReleased.compareAndSet(false, true)) {
```

Comment:
Why atomic? It could easily be checked and set under `synchronized (receivedBuffers)` lock, simplifying threading model a bit and it would also avoid extra `AtomicBoolean` check on the hot path.

Reply:
In `RemoteInputChannel` we also had the atomic released variable and I guess here is the similar case. The release operation might be called not only by task thread, but also by canceler thread. If the latter, it might has visibility issue for task thread while interacting with `RecoveredInputChannel#isReleased` in `BufferManager` related operations.

Reply:
> also by canceler thread. If the latter, it might has visibility issue 

it wouldn't:

>  It could easily be checked and set under `synchronized (receivedBuffers)` lock

?

Reply:
yeah, we can also take the way of adding the `synchronized (receivedBuffers)` for the  method of `RecoveredInputGate#isReleased()`

Reply:
It seems we have different ways for this issue in different classes. `LocalInputChannel` introduces `volatile` for this variable, and `RemoteInputChannel` introduces `atomic` variable, but `RecoveredInputChannel` takes the normal boolean variable. 

We might need to unify them in separate commits future.

Reply:
Agreed. But generally speaking closing/releasing logic is on my list for a larger refactor/clean up for a long time, especially because of those various nasty concurrency issues - I would see closing/releasing eventually moved to task thread/mailbox.